app/controllers/sites_controller.rb

```diff
@@ -1,8 +1,62 @@
 class SitesController < ApplicationController
-  before_action :authenticate
+  before_action :authenticate, :build_provider, :initialise_errors
 
   def index
-    @provider = Provider.includes(:sites).find(params[:provider_code]).first
     @sites = @provider.sites.sort_by(&:location_name)
   end
+
+  def edit
+    @site = @provider.sites.find { |site| site.id == params[:id] }
+  end
+
+  def update
+    @site = @provider.sites.find { |site| site.id == params[:id] }
+
+    # We don't have a provider_code in the backend, only provider_id, and this
+    # is required in order to run #update.
+    @site.provider_code = @provider.provider_code
+
+    if @site.update(site_params)
+      redirect_to provider_sites_path, flash: { success: 'Your changes have been published' }
+    else
+      @errors = @site.errors.reduce({}) { |errors, (field, message)|
+        errors[field] ||= []
+        errors[field].push(map_errors(message))
+        errors
+      }
```

Comment:
Let's not forget to refactor this into a method as we'll want to re-use it. Also, is this a cleaner implementation?

```
@errors = @site.errors
  .group_by { |field, _message| field }
  .transform_values { |group| group.map &:second }
```

Maybe I'm blinded by a slight aversion to using `reduce`, just a thought.

Reply:
I think the reducer is actually easier to understand, I had to `binding.pry` to figure that one out. :)